File: go-flood/pkg/tilings/basket_weave.go
package tilings

import (
	"go-flood/pkg/core"
	"math"
	"sort"
)

type basketWeaveRawTile struct {
	weaveType string
	ix        int
	iy        int
	w         int
	h         int
}

// GenerateBasketWeaveBoard generates a basket-weave tiling from congruent 3x1
// rectangles in alternating orientations, with 1x1 square tiles in the gaps.
func GenerateBasketWeaveBoard(options Options) core.Board {
	pad := 4
	rawTiles := []basketWeaveRawTile{}
	seen := map[[4]int]bool{}

	mod := func(n, m int) int {
		return ((n % m) + m) % m
	}

	intersectsTarget := func(weaveType string, ix, iy, w, h int) bool {
		overlapsTarget := ix < options.Cols && ix+w > 0 && iy < options.Rows && iy+h > 0
		if overlapsTarget {
			return true
		}

		// Balance the natural top/right protrusions by also keeping the matching
		// vertical strips that touch both side edges and horizontal strips that
		// touch the bottom edge.
		touchesLeftEdge := weaveType == "vertical" && ix+w == 0 && iy < options.Rows && iy+h > 0
		touchesRightEdge := weaveType == "vertical" && ix == options.Cols && iy < options.Rows && iy+h > 0
		touchesBottomEdge := weaveType == "horizontal" && iy == options.Rows && ix < options.Cols && ix+w > 0
		return touchesLeftEdge || touchesRightEdge || touchesBottomEdge
	}

	addTile := func(weaveType string, ix, iy, w, h int) {
		if !intersectsTarget(weaveType, ix, iy, w, h) {
			return
		}
		key := [4]int{ix, iy, w, h}
		if seen[key] {
			return
		}
		seen[key] = true
		rawTiles = append(rawTiles, basketWeaveRawTile{weaveType: weaveType, ix: ix, iy: iy, w: w, h: h})
	}

	minX := -pad
	maxX := options.Cols + pad
	minY := -pad
	maxY := options.Rows + pad

	// Horizontal 3x1 rectangles: even rows, staggered by phase modulo 4.
	for iy := minY; iy <= maxY; iy++ {
		if mod(iy, 2) != 0 {
			continue
		}
		for ix := minX; ix <= maxX; ix++ {
			if mod(ix, 4) == mod(iy, 4) {
				addTile("horizontal", ix, iy, 3, 1)
			}
		}
	}

	// Vertical 1x3 rectangles: odd columns, staggered by phase modulo 4.
	for ix := minX; ix <= maxX; ix++ {
		if mod(ix, 2) != 1 {
			continue
		}
		for iy := minY; iy <= maxY; iy++ {
			if mod(iy, 4) == mod(ix, 4) {
				addTile("vertical", ix, iy, 1, 3)
			}
		}
	}

	// 1x1 squares sit on odd rows and even columns.
	for iy := minY; iy <= maxY; iy++ {
		if mod(iy, 2) != 1 {
			continue
		}
		for ix := minX; ix <= maxX; ix++ {
			if mod(ix, 2) == 0 {
				addTile("square", ix, iy, 1, 1)
			}
		}
	}

	sort.Slice(rawTiles, func(i, j int) bool {
		ai := rawTiles[i]
		aj := rawTiles[j]
		cyI := float64(ai.iy) + float64(ai.h)/2
		cyJ := float64(aj.iy) + float64(aj.h)/2
		if cyI != cyJ {
			return cyI < cyJ
		}
		cxI := float64(ai.ix) + float64(ai.w)/2
		cxJ := float64(aj.ix) + float64(aj.w)/2
		return cxI < cxJ
	})

	tiles := make([]core.Tile, len(rawTiles))
	for id, raw := range rawTiles {
		tiles[id] = core.Tile{
			ID:        id,
			ColorID:   int(options.RNG() * float64(options.ColorCount)),
			OwnerID:   nil,
			Points:    basketWeaveRectanglePoints(float64(raw.ix)*options.TileSize, float64(raw.iy)*options.TileSize, raw.w, raw.h, options.TileSize),
			Neighbors: []int{},
		}
	}

	cellToTile := map[[2]int]int{}
	for tileID, raw := range rawTiles {
		for dy := 0; dy < raw.h; dy++ {
			for dx := 0; dx < raw.w; dx++ {
				cellToTile[[2]int{raw.ix + dx, raw.iy + dy}] = tileID
			}
		}
	}

	neighborSets := make([]map[int]bool, len(tiles))
	for i := range neighborSets {
		neighborSets[i] = map[int]bool{}
	}

	for cell, tileID := range cellToTile {
		candidates := [][2]int{{cell[0] + 1, cell[1]}, {cell[0], cell[1] + 1}}
		for _, candidate := range candidates {
			neighborID, ok := cellToTile[candidate]
			if ok && neighborID != tileID {
				neighborSets[tileID][neighborID] = true
				neighborSets[neighborID][tileID] = true
			}
		}
	}

	for id := range tiles {
		neighbors := make([]int, 0, len(neighborSets[id]))
		for neighborID := range neighborSets[id] {
			neighbors = append(neighbors, neighborID)
		}
		sort.Ints(neighbors)
		tiles[id].Neighbors = neighbors
	}

	minPX, minPY, _, _ := basketWeaveBoundingBox(tiles)
	for i := range tiles {
		for j, p := range tiles[i].Points {
			tiles[i].Points[j] = core.Point{p[0] - minPX, p[1] - minPY}
		}
	}

	_, _, width, height := basketWeaveBoundingBox(tiles)
	startTileIds := basketWeaveUniqueIds([]int{
		basketWeaveClosestTileID(tiles, 0, 0),
		basketWeaveClosestTileID(tiles, width, height),
		basketWeaveClosestTileID(tiles, width, 0),
		basketWeaveClosestTileID(tiles, 0, height),
	})

	return core.Board{
		Version:      1,
		Generator:    "basket-weave",
		Width:        width,
		Height:       height,
		Cols:         options.Cols,
		Rows:         options.Rows,
		Tiles:        tiles,
		StartTileIds: startTileIds,
	}
}

func basketWeaveRectanglePoints(x, y float64, wCells, hCells int, unit float64) []core.Point {
	points := []core.Point{}

	for i := 0; i <= wCells; i++ {
		points = append(points, core.Point{x + float64(i)*unit, y})
	}
	for j := 1; j <= hCells; j++ {
		points = append(points, core.Point{x + float64(wCells)*unit, y + float64(j)*unit})
	}
	for i := wCells - 1; i >= 0; i-- {
		points = append(points, core.Point{x + float64(i)*unit, y + float64(hCells)*unit})
	}
	for j := hCells - 1; j >= 1; j-- {
		points = append(points, core.Point{x, y + float64(j)*unit})
	}

	return points
}

func basketWeaveBoundingBox(tiles []core.Tile) (float64, float64, float64, float64) {
	minX := math.MaxFloat64
	minY := math.MaxFloat64
	maxX := -math.MaxFloat64
	maxY := -math.MaxFloat64

	for _, tile := range tiles {
		for _, p := range tile.Points {
			minX = math.Min(minX, p[0])
			minY = math.Min(minY, p[1])
			maxX = math.Max(maxX, p[0])
			maxY = math.Max(maxY, p[1])
		}
	}

	return minX, minY, maxX, maxY
}

func basketWeaveClosestTileID(tiles []core.Tile, tx, ty float64) int {
	bestID := -1
	bestDist := math.MaxFloat64

	for _, tile := range tiles {
		cx, cy := basketWeaveCentroid(tile.Points)
		dist := math.Hypot(cx-tx, cy-ty)
		if dist < bestDist {
			bestDist = dist
			bestID = tile.ID
		}
	}

	return bestID
}

func basketWeaveCentroid(points []core.Point) (float64, float64) {
	x := 0.0
	y := 0.0
	for _, p := range points {
		x += p[0]
		y += p[1]
	}
	return x / float64(len(points)), y / float64(len(points))
}

func basketWeaveUniqueIds(ids []int) []int {
	seen := map[int]bool{}
	unique := []int{}
	for _, id := range ids {
		if id == -1 || seen[id] {
			continue
		}
		seen[id] = true
		unique = append(unique, id)
	}
	return unique
}

File: js/tilings/basket-weave.js
/**
 * Generates a basket-weave tiling board as a graph.
 *
 * The pattern is built on a square unit grid from congruent 3x1 rectangles
 * in alternating horizontal/vertical orientations, with 1x1 square tiles in
 * the gaps.  Rectangles are emitted with collinear perimeter vertices at unit
 * intervals so renderer edge maps can match the smaller neighboring edges.
 */
export function generateBasketWeaveBoard(options) {
  const { cols, rows, tileSize, colorCount, rng } = options;
  const unit = tileSize;
  const rawTiles = [];
  const seen = new Set();
  const pad = 4;

  const mod = (n, m) => ((n % m) + m) % m;

  const intersectsTarget = (weaveType, ix, iy, w, h) => {
    const overlapsTarget = ix < cols && ix + w > 0 && iy < rows && iy + h > 0;
    if (overlapsTarget) return true;

    // Balance the natural top/right protrusions by also keeping the matching
    // vertical strips that touch both side edges and horizontal strips that
    // touch the bottom edge.
    const touchesLeftEdge = weaveType === 'vertical' && ix + w === 0 && iy < rows && iy + h > 0;
    const touchesRightEdge = weaveType === 'vertical' && ix === cols && iy < rows && iy + h > 0;
    const touchesBottomEdge = weaveType === 'horizontal' && iy === rows && ix < cols && ix + w > 0;
    return touchesLeftEdge || touchesRightEdge || touchesBottomEdge;
  };

  const addTile = (weaveType, ix, iy, w, h) => {
    if (!intersectsTarget(weaveType, ix, iy, w, h)) return;
    const key = `${ix},${iy},${w},${h}`;
    if (seen.has(key)) return;
    seen.add(key);
    rawTiles.push({ weaveType, ix, iy, w, h });
  };

  const minX = -pad;
  const maxX = cols + pad;
  const minY = -pad;
  const maxY = rows + pad;

  // Horizontal 3x1 rectangles: even rows, staggered by phase modulo 4.
  for (let iy = minY; iy <= maxY; iy++) {
    if (mod(iy, 2) !== 0) continue;
    for (let ix = minX; ix <= maxX; ix++) {
      if (mod(ix, 4) === mod(iy, 4)) {
        addTile('horizontal', ix, iy, 3, 1);
      }
    }
  }

  // Vertical 1x3 rectangles: odd columns, staggered by phase modulo 4.
  for (let ix = minX; ix <= maxX; ix++) {
    if (mod(ix, 2) !== 1) continue;
    for (let iy = minY; iy <= maxY; iy++) {
      if (mod(iy, 4) === mod(ix, 4)) {
        addTile('vertical', ix, iy, 1, 3);
      }
    }
  }

  // 1x1 squares sit on odd rows and even columns.
  for (let iy = minY; iy <= maxY; iy++) {
    if (mod(iy, 2) !== 1) continue;
    for (let ix = minX; ix <= maxX; ix++) {
      if (mod(ix, 2) === 0) {
        addTile('square', ix, iy, 1, 1);
      }
    }
  }

  // Stable spatial order keeps tile IDs predictable and start positions sane.
  rawTiles.sort((a, b) => {
    const ay = a.iy + a.h / 2;
    const by = b.iy + b.h / 2;
    if (ay !== by) return ay - by;
    const ax = a.ix + a.w / 2;
    const bx = b.ix + b.w / 2;
    return ax - bx;
  });

  const tiles = rawTiles.map((raw, id) => ({
    id,
    colorId: Math.floor(rng() * colorCount),
    ownerId: null,
    points: rectanglePoints(raw.ix * unit, raw.iy * unit, raw.w, raw.h, unit),
    neighbors: [],
    weaveType: raw.weaveType
  }));

  const cellToTile = new Map();
  rawTiles.forEach((tile, tileId) => {
    for (let dy = 0; dy < tile.h; dy++) {
      for (let dx = 0; dx < tile.w; dx++) {
        cellToTile.set(`${tile.ix + dx},${tile.iy + dy}`, tileId);
      }
    }
  });

  const neighborSets = tiles.map(() => new Set());
  for (const [key, tileId] of cellToTile.entries()) {
    const [x, y] = key.split(',').map(Number);
    for (const [nx, ny] of [[x + 1, y], [x, y + 1]]) {
      const neighborId = cellToTile.get(`${nx},${ny}`);
      if (neighborId !== undefined && neighborId !== tileId) {
        neighborSets[tileId].add(neighborId);
        neighborSets[neighborId].add(tileId);
      }
    }
  }

  tiles.forEach((tile, id) => {
    tile.neighbors = [...neighborSets[id]].sort((a, b) => a - b);
  });

  let bbox = computeBoundingBox(tiles);
  for (const tile of tiles) {
    tile.points = tile.points.map(([x, y]) => [x - bbox.minX, y - bbox.minY]);
  }
  bbox = computeBoundingBox(tiles);

  const startTileIds = uniqueIds([
    closestTileId(tiles, 0, 0),
    closestTileId(tiles, bbox.maxX, bbox.maxY),
    closestTileId(tiles, bbox.maxX, 0),
    closestTileId(tiles, 0, bbox.maxY)
  ]);

  return {
    version: 1,
    generator: 'basket-weave',
    width: bbox.maxX,
    height: bbox.maxY,
    cols,
    rows,
    tiles,
    startTileIds
  };
}

function rectanglePoints(x, y, wCells, hCells, unit) {
  const points = [];

  for (let i = 0; i <= wCells; i++) points.push([x + i * unit, y]);
  for (let j = 1; j <= hCells; j++) points.push([x + wCells * unit, y + j * unit]);
  for (let i = wCells - 1; i >= 0; i--) points.push([x + i * unit, y + hCells * unit]);
  for (let j = hCells - 1; j >= 1; j--) points.push([x, y + j * unit]);

  return points;
}

function computeBoundingBox(tiles) {
  let minX = Infinity, minY = Infinity;
  let maxX = -Infinity, maxY = -Infinity;

  for (const tile of tiles) {
    for (const [x, y] of tile.points) {
      minX = Math.min(minX, x);
      minY = Math.min(minY, y);
      maxX = Math.max(maxX, x);
      maxY = Math.max(maxY, y);
    }
  }

  return { minX, minY, maxX, maxY };
}

function closestTileId(tiles, tx, ty) {
  let bestId = -1;
  let bestDist = Infinity;

  for (const tile of tiles) {
    const [cx, cy] = centroid(tile.points);
    const dist = Math.hypot(cx - tx, cy - ty);
    if (dist < bestDist) {
      bestDist = dist;
      bestId = tile.id;
    }
  }

  return bestId;
}

function centroid(points) {
  let x = 0, y = 0;
  for (const [px, py] of points) {
    x += px;
    y += py;
  }
  return [x / points.length, y / points.length];
}

function uniqueIds(ids) {
  return [...new Set(ids)].filter(id => id !== -1);
}
